### PR TITLE
Create a private channel one week before the birthday

### LINKED ISSUE #87: Create a private channel one week before the birthday
A private group with all users except the birthday boy/girl must be created one week before the birthday.

### PR DESCRIPTION
**Pull request by ya7on**
close #87

### CODE REVIEW

File: src/birthder.js [hunk 1]
```diff
@@ -68,6 +68,41 @@
...
+   * @param {string} username Birthday boy
+   * @returns {Void}
+   */
+  const createBirthdayRoom = async (robot, username) => {
+    const message = `Soon our colleague @${username} celebrates his birthday. Let's discuss this`
+    const roomName = `${username}_happy_birthday_room`
+    const users = Object.values(robot.brain.data.users)
+      .filter(user => user.name !== username)
+      .map(user => user.name)
+    console.log('JSON' + JSON.stringify(users))
```
Comment (eugulixes) [inline src/birthder.js]: Get rid of the debugging log.

File: src/birthder.js [hunk 2]
```diff
@@ -68,6 +68,41 @@
...
+    const room = await robot.adapter.api.post('groups.create', {
```
Comment (eugulixes) [inline src/birthder.js:83]: Is it a private group?

File: src/birthder.js [hunk 3]
```diff
@@ -68,6 +68,41 @@
     }
   }
 
+  /** 
+   * Creating a room for all users except `username`
+   * @param {Robot} robot Hubot instance
+   * @param {string} username Birthday boy
```
Comment (eugulixes) [inline src/birthder.js]: Birthday boy/girl -- let's be politically correct.

File: src/birthder.js [hunk 4]
```diff
@@ -68,6 +68,41 @@
     }
   }
 
+  /** 
+   * Creating a room for all users except `username`
```
Comment (eugulixes) [inline src/birthder.js]: Use imperative form of the verb here, i.e. `Create`.

File: src/birthder.js [hunk 5]
```diff
@@ -68,6 +68,41 @@
...
+  }
+
+  /** 
+   * Check if bot is in the happy birthday group
+   * @param {Robot} robot Hubot instance
+   * @param {string} username birthday boy's username
+   * @returns {boolean}
+   */
+  const isBotInHBGroup = async (robot, username) => {
+    const roomName = `${username}_happy_birthday_room`
```
Comment (eugulixes) [inline src/birthder.js]: * Let's name the variable `channelName`.
* Let's name the group `${username}-birthday-channel`.

File: src/birthder.js (hunk 2 above)
Comment (ya7on) [reply src/birthder.js:83]: > Is it a private group?

Yes. A private channel in the API is called "group"

**Comment by eugulixes** [comment]
Why the whole file `package.json` is marked as changed?

File: src/birthder.js [hunk 6]
```diff
@@ -226,6 +283,11 @@
     message = formReminderMessage(users, targetDay, amountOfTime)
 
     if (users.length > 0) {
+      for (let birthdayboy of users) {
+        if (!await isBotInHBGroup(robot, birthdayboy.name)) {
+          await createBirthdayRoom(robot, birthdayboy.name)
+        }
+      }
```
Comment (eugulixes) [inline src/birthder.js:290]: Put a new line right after the brace.

File: src/birthder.js [hunk 7]
```diff
@@ -68,6 +68,40 @@
     }
   }
 
+  /**
+   * Create a room for all users except `username`
+   * @param {Robot} robot Hubot instance
+   * @param {string} username Birthday boy/girl
+   * @returns {Void}
+   */
+  const createBirthdayRoom = async (robot, username) => {
```
Comment (eugulixes) [inline src/birthder.js]: The function creates `${username}-birthday-channel`, so name it `createBirthdayChannel`.

File: src/birthder.js [hunk 8]
```diff
@@ -68,6 +68,40 @@
     }
   }
 
+  /**
+   * Create a room for all users except `username`
```
Comment (eugulixes) [inline src/birthder.js]: Create a channel and invite all the users to it except the one specified via `username`.

File: src/birthder.js [hunk 9]
```diff
@@ -209,12 +243,35 @@
     return message
   }
 
+  /**
+   * Check if there is a birthday yesterday
+   * @param {Robot} robot Hubot instance
+   * @returns {Void}
+   */
+  async function checkYesterday (robot) {
```
Comment (eugulixes) [inline src/birthder.js]: I guess the name `removeExpiredBirthdayChannels` would reflect the purpose of the function better.

File: src/birthder.js [hunk 10]
```diff
@@ -68,6 +68,40 @@
...
+      name: channelName,
+      members: users
+    })
+    robot.messageRoom(room.group.name, message)
+  }
+
+  /**
+   * Check if bot is in the happy birthday group
+   * @param {Robot} robot Hubot instance
+   * @param {string} username birthday boy/girl's username
```
Comment (eugulixes) [inline src/birthder.js]: Let's substitute it for `Username of birthday boy/girl`

File: src/birthder.js [hunk 11]
```diff
@@ -68,6 +68,40 @@
...
+      .map(user => user.name)
+    const room = await robot.adapter.api.post('groups.create', {
+      name: channelName,
+      members: users
+    })
+    robot.messageRoom(room.group.name, message)
+  }
+
+  /**
+   * Check if bot is in the happy birthday group
```
Comment (eugulixes) [inline src/birthder.js]: Check if the bot is in the birthday channel

File: src/birthder.js [hunk 12]
```diff
@@ -68,6 +68,40 @@
...
   }
 
+  /**
+   * Create a room for all users except `username`
+   * @param {Robot} robot Hubot instance
+   * @param {string} username Birthday boy/girl
+   * @returns {Void}
+   */
+  const createBirthdayRoom = async (robot, username) => {
+    const message = `Soon our colleague @${username} celebrates his birthday. Let's discuss this`
```
Comment (eugulixes) [inline src/birthder.js]: @${username} is having a birthday soon, so let's discuss a present.

File: src/birthder.js [hunk 13]
```diff
@@ -68,6 +68,40 @@
...
+    robot.messageRoom(room.group.name, message)
+  }
+
+  /**
+   * Check if bot is in the happy birthday group
+   * @param {Robot} robot Hubot instance
+   * @param {string} username birthday boy/girl's username
+   * @returns {boolean}
+   */
+  const isBotInHBGroup = async (robot, username) => {
```
Comment (eugulixes) [inline src/birthder.js]: `isBotInBirthdayChannel` would be better I guess.

File: src/birthder.js [hunk 14]
```diff
@@ -209,12 +243,35 @@
     return message
   }
 
+  /**
+   * Check if there is a birthday yesterday
```
Comment (eugulixes) [inline src/birthder.js]: Find all the birthdays which were yesterday and remove the channels created for them.

**Comment by eugulixes** [comment]
Add something like this to the Features list:
```Creates private channels in advance and invites all the users to the channels except the birthday boys/girls for the purpose of discussing the presents for them. Expired channels are automatically removed. ```

File: src/birthder.js (hunk 2 above)
Comment (eugulixes) [reply src/birthder.js:83]: Ok, thank you.

File: src/birthder.js [hunk 15]
```diff
@@ -68,6 +68,40 @@
...
+      name: channelName,
+      members: users
+    })
+    robot.messageRoom(room.group.name, message)
+  }
+
+  /**
+   * Check if the bot is in the birthday channel
+   * @param {Robot} robot Hubot instance
+   * @param {string} username Username
```
Comment (eugulixes) [inline src/birthder.js]: Username of birthday boy/girl

File: src/birthder.js [hunk 16]
```diff
@@ -68,6 +68,40 @@
     }
   }
 
+  /**
+   * Create a channel and invite all the users to it except the one specified via username.
+   * @param {Robot} robot Hubot instance
+   * @param {string} username Username
```
Comment (eugulixes) [inline src/birthder.js]: Username of birthday boy/girl

**Review by polina-popova — APPROVED** [review]

**Pull request merged**